### PR TITLE
feat: add chisel as a build snap when stage-packages contain slices

### PR DESCRIPTION
**Pull request by linostar**
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

This PR aims to stop chisel from being unnecessarily installed in the Rockcraft snap.

It does that by having chisel as a build snap only when the following conditions are met:
- there is at least one `part` that has slices in its `stage-packages`
- there is no `part` that has chisel in its `build-snaps`

### CODE REVIEW

File: craft_parts/parts.py [hunk 1]
```diff
@@ -175,6 +175,18 @@ def has_overlay(self) -> bool:
...
 
+    @property
+    def has_slices(self) -> bool:
+        """Return whether the part contains chisel slices."""
+        return any(p for p in self.stage_packages if "_" in p)
+
+    @property
+    def has_chisel_as_build_snap(self) -> bool:
+        """Return whether the part has chisel as build snap."""
+        return any(
```
Comment (SamirPS) [inline craft_parts/parts.py:186]: can't we just do 

` p for p in self.build_snaps if p.startswith("chisel")`

Comment (linostar) [reply craft_parts/parts.py:190]: No, because if there was a snap called "chiselhelper" for instance, it will match with that, which we don't want.

Comment (SamirPS) [reply craft_parts/parts.py:190]: yeah true

**Comment by linostar** [comment]
@canonical/starcraft-reviewers This is ready for review.

File: tests/unit/test_parts.py [hunk 2]
```diff
@@ -335,6 +335,31 @@ def test_part_has_overlay(self, partitions):
...
+        p = Part("foo", tc_spec, partitions=partitions)
+        assert p.spec.has_slices == tc_result
+
+    @pytest.mark.parametrize(
+        ("tc_spec", "tc_result"),
+        [
+            ({}, False),
+            ({"build-snaps": ["hello", "chisel/latest/candidate"]}, True),
+            ({"build-snaps": ["chisel"]}, True),
+            ({"build-snaps": ["hello"]}, False),
```
Comment (mr-cal) [inline tests/unit/test_parts.py:356]: You had a good point in your comment, I think you should test it.

```suggestion
            ({"build-snaps": ["hello"]}, False),
            ({"build-snaps": ["chiselhelper"]}, False),
```

**Review by mr-cal — APPROVED** [review]

File: tests/unit/test_parts.py (hunk 2 above)
Comment (linostar) [reply tests/unit/test_parts.py:357]: Added.

**Pull request merged**